### PR TITLE
4762: Ding mobilesearch paragraph export to CMS content

### PR DESCRIPTION
**Pull request by holt83**
#### Link to issue

https://platform.dandigbib.org/issues/4762#note-23 (follow-up issue)

#### Description

Paragraphs are now also added to the export mapping in
ding_mobilesearch export hook implementation.

Also move the field data extraction code for common fields to a
separate function and use that for both node and paragraphs_item
entities. It was the easiest way to reuse all the special handling for
each wiget type.

#### Example of ding_new export with paragraphs items:

```
"id": "6007741d841d44231cf93ae3",
"nid": 41,
"agency": "763000",
"type": "ding_news",
"fields": {
  "ding_news_groups_ref": {
    "name": "Groups",
    "value": [
      "Faglitteratur - uhmm!"
    ],
    "attr": []
  },
  "field_ding_news_lead": {
    "name": "Lead",
    "value": "Curabitur eu nibh ac purus sagittis molestie consectetur nec dolor. Nunc sed scelerisque dolor. LOL",
    "attr": []
  },
  "field_ding_news_list_image": {
    "name": "List image",
    "value": [
      "files\/763000\/original\/37f0358a670175b387b0bd0189fc24a0ed6526bc.jpeg"
    ],
    "attr": [
      "image\/jpeg"
    ]
  },
  "field_ding_news_materials": {
    "name": "Materials",
    "value": {
      "materials": [
        "870970-basis:47935024"
      ]
    },
    "attr": []
  },
  "og_group_ref": {
    "name": "Library",
    "value": [
      "Lokalbiblioteket"
    ],
    "attr": []
  },
  "title": {
    "name": "Titel",
    "value": "Sportsstjerner - bag facaden",
    "attr": []
  },
  "author": {
    "name": "Forfatter",
    "value": "admin",
    "attr": []
  },
  "created": {
    "name": "Oprettet",
    "value": "2013-12-16T22:37:52+01:00",
    "attr": []
  },
  "changed": {
    "name": "Opdateret",
    "value": "2021-01-20T01:05:07+01:00",
    "attr": []
  },
  "status": {
    "name": "Status",
    "value": 1,
    "attr": []
  },
  "field_ding_news_tags": {
    "name": "Tags",
    "value": [
      "fodbold",
      "sportsstjerner"
    ],
    "attr": []
  },
  "field_ding_news_paragraphs": {
    "name": "Paragraphs",
    "value": [
      {
        "type": "ding_paragraphs_text",
        "fields": {
          "field_ding_paragraphs_text": {
            "name": "Text",
            "value": "<p>En lille test..<\/p>",
            "attr": []
          }
        }
      },
      {
        "type": "ding_paragraphs_material_list",
        "fields": {
          "field_ding_paragraphs_material": {
            "name": "Materiale",
            "value": {
              "materials": [
                "870970-basis:47935024"
              ]
            },
            "attr": []
          }
        }
      },
      {
        "type": "ding_paragraphs_image",
        "fields": {
          "field_ding_paragraphs_display": {
            "name": "Show as",
            "value": [
              "half_right"
            ],
            "attr": []
          },
          "field_ding_paragraphs_image": {
            "name": "Image",
            "value": [
              "\/9j\/4AAQSkZJRgABAQEASABIAAD\/4RU2RXhpZgAASUk..."
            ],
            "attr": [
              "image\/jpeg"
            ]
          }
        }
      },
      {
        "type": "ding_paragraphs_carousel",
        "fields": {
          "field_ding_paragraphs_material": {
            "name": "Materiale",
            "value": {
              "materials": [
                "870970-basis:47935024",
                "870970-basis:38609734"
              ]
            },
            "attr": []
          }
        }
      },
      {
        "type": "ding_paragraphs_image_and_text",
        "fields": {
          "field_ding_paragraphs_display": {
            "name": "Show as",
            "value": [
              "full_width"
            ],
            "attr": []
          },
          "field_ding_paragraphs_image": {
            "name": "Image",
            "value": [
              "\/9j\/4AAQSkZJRgABAQAAAQABAAD\/\/gA7Q1JFQVRPUjo..."
            ],
            "attr": [
              "image\/jpeg"
            ]
          },
          "field_ding_paragraphs_position": {
            "name": "Image position",
            "value": [
              "ting_object_left"
            ],
            "attr": []
          },
          "field_ding_paragraphs_text": {
            "name": "Text",
            "value": "<p>Image text<\/p>",
            "attr": []
          }
        }
      },
      {
        "type": "ding_paragraphs_single_material",
        "fields": {
          "field_ding_paragraphs_display": {
            "name": "Show as",
            "value": [
              "half_left"
            ],
            "attr": []
          },
          "field_ding_paragraphs_single_mat": {
            "name": "Materiale",
            "value": {
              "materials": [
                "870970-basis:47935024"
              ]
            },
            "attr": []
          }
        }
      },
      {
        "type": "ding_paragraphs_text_box",
        "fields": {
          "field_ding_paragraphs_box_text": {
            "name": "Box text",
            "value": "<p>Text box paragraph test<\/p>",
            "attr": []
          },
          "field_ding_paragraphs_display": {
            "name": "Show as",
            "value": [
              "full_width"
            ],
            "attr": []
          }
        }
      }
    ],
    "attr": []
  }
},
"taxonomy": {
  "field_ding_news_category": {
    "name": "News Category",
    "terms": {
      "Film": []
    }
  }
},
"list": []
```

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

### CODE REVIEW

File: modules/ding_mobilesearch/ding_mobilesearch.module [hunk 1]
```diff
@@ -117,61 +117,174 @@ function _ding_mobilesearch_types() {
...
  */
 function ding_mobilesearch_ding_mobilesearch_node_export_mapping($node) {
+  // Map common fields on the node.
+  $mapping = _ding_mobilesearch_map_entity_fields('node', $node->type, $node);
+
   $username = db_select('users', 'u')
     ->fields('u', array('name'))
     ->condition('u.uid', $node->uid)
     ->execute()
     ->fetchField();
```
Comment (kasperg) [inline modules/ding_mobilesearch/ding_mobilesearch.module:131]: This is not really a the proper way to get the username for the author of a node but I guess that is not part of the scope of this PR 😕 .

File: modules/ding_mobilesearch/ding_mobilesearch.module [hunk 2]
```diff
@@ -117,61 +117,174 @@ function _ding_mobilesearch_types() {
...
+    }, $paragraphs_item_entities);
+  }
+
+  return $mapping;
+}
+
+/**
+ * Helper function to map common fields from an entity.
+ */
+function _ding_mobilesearch_map_entity_fields($entity_type, $bundle, $entity) {
```
Comment (kasperg) [inline modules/ding_mobilesearch/ding_mobilesearch.module:271]: :+1: but remember to document function arguments.

**Review by kasperg — APPROVED** [review]
:+1: LGTM

File: modules/ding_mobilesearch/ding_mobilesearch.module (hunk 1 above)
Comment (holt83) [reply modules/ding_mobilesearch/ding_mobilesearch.module:131]: Tbh this was a case of getting in and out as fast as possible 😄 Not really a fan of working with this module in general.

But I'll happily update this. Using  `format_username()` on a loaded user object should give more correct result.

File: modules/ding_mobilesearch/ding_mobilesearch.module (hunk 2 above)
Comment (holt83) [reply modules/ding_mobilesearch/ding_mobilesearch.module]: Added documentation.

File: modules/ding_mobilesearch/ding_mobilesearch.module (hunk 1 above)
Comment (holt83) [reply modules/ding_mobilesearch/ding_mobilesearch.module]: Done.

**Pull request merged**